### PR TITLE
Fast logger migration

### PR DESCRIPTION
**Pull request by RiugaBachi**
This PR implements the changes discussed in #257 as well as some concerns discussed privately, namely:

- Configurable logging to stderr (via `rotate-logs` config option)
  - This can be useful i.e. when keter is managed via a `systemd`, which automatically captures and tails stderr output.
- Implementation of a `KeterM cfg a` monad to encapsulate a (logging + config) context, which crops up prominently.
  - For contexts that _only_ require the logging capability, we can simply leave `cfg` polymorphic!
- Refactoring of relevant modules to use the new KeterM monad to:
  - Allow access to `MonadLogger` -related capabilities via the internal `LoggingT` representation.
  - Cut down on redundant passing of the same configs and loggers everywhere.
- Changing of logging calls to use TH-based logging from `monad-logger`, giving us access to `Loc` (module+line number) info at each call site.
- Reformatting of log messages to make use of the new info with format: `"keter|$time|$module$:$line_num|$log_level> $msg` 
- Removal of the `LogMessage` ADT and its gigantic `Show` instance, which was a bit overkill and clunky for logging.
- Removal of the old in-house logging solution.
- Deprecation of the old `logEx` utility.

As was discussed privately, log rotation naming and directory scheme have had to be changed due to limitations in `fast-logger`. Namely, that it dictates the naming scheme of rotated logs with no straightforward way of customizing it to preserve backwards compatibility. On top of the public keter API changes due to the `KeterM` refactor, package version has been bumped to `2.1.0` to indicate a breaking change.

Relevant parts of the test suite have also been rewritten to comply with the API changes, but overall semantics are still retained.

---

TODO:

- Resolve individual todo comments (@jappeace could use your input on some assumptions :pray:)
- Version bounds on monad-logger / fast-logger / unliftio-core?
- ~~Refactoring~~
  * ~~Possibly some things we can move to LogFile module~~
  * ~~Rename/change dir of LogFile module?~~
  * ~~Investigate if we really still need a TVar for RotatingLog since `fast-logger` was implemented with the assumption of multithreaded contexts.~~
  * ~~Formatting~~
 - Propagate a logging context around :heavy_check_mark: 
 -  Change individual log calls to source line log splices based on the show instance for the log message ADT. :heavy_check_mark: 
 - Then remove said ADT. :heavy_check_mark: 
- Fix stubbed out parts of the test suite :heavy_check_mark: 
 - Update changelog again :heavy_check_mark: 
 - Bump version in cabal file :heavy_check_mark:

### CODE REVIEW

File: ChangeLog.md [hunk 1]
```diff
@@ -1,3 +1,25 @@
...
+
++ Log naming and directory scheme has changed for both main keter logs and app logs.  
+  Old logs were named `dir/current.log` for the current log and `%Y%m%d_%H%M%S.log` 
+  (`time` package conventions) for rotated logs.  
+  Current logs have been brought up one level and named after their old directory:  
+  `logs/keter/current.log` -> `logs/keter.log`  
+  Rotated logs will now simply have `.1` `.2` ascending appended to the name of the base logs 
+  rather than be named after the date and time they were rotated at:  
+  `logs/keter/20230413_231415.log` -> `logs/keter/keter.log.1`  
+  Please update anything that depended on the old log naming and directory conventions accordingly.
```
Comment (jappeace) [inline ChangeLog.md:11]: how does this work with several apps?

Comment (RiugaBachi) [reply ChangeLog.md:13]: Noticed a typo here actually :pray: Added some examples for app log changes just now.

**Comment by jappeace** [comment]
I think deleting that log message datatype would be nice

https://github.com/snoyberg/keter/blob/1bd7e2591906c0e719affeb48e0b7c3a415650d4/src/Keter/Common.hs#L59

I've no idea why this was made typesafe, I hardly can imagine a situation you want to parse and re-use log messages.

----

anyway, like I said you don't need an instance because you can use LoggingT: https://hackage.haskell.org/package/monad-logger-0.3.39/docs/Control-Monad-Logger.html#t:LoggingT

fast logger provides you with `LogStr -> IO ()`, so you just have to tell it how to deal with those other args and `Loc -> LogSource -> LogLevel` (eg format them into a LogStr for example and concat the given LogStr, or just ignore them).
Then we can use:

```haskell
runLoggingT (myFormatting fastlogHandle) $ $logInfo "messages from LogMessage as normal text"
```

**Comment by RiugaBachi** [comment]
Oh lol my eyes completely skipped over the unwrapper being exposed. Thanks for the heads up :+1: 

Though, I was considering my options and still think the best path forward would be to propagate a new `KeterM` monad context to everything. This would allow for extensibility if at some point we decide to throw in a `ReaderT KeterConfig` stack etc. for other resources that seem to be passed around to everything besides the logger. I couldn't see a benefit to generalizing everything to MonadLogger/IO/etc. constraints since keter (main) stuff forms a pretty linear call graph and the polymorphism wouldn't be useful in practice.

**Review by RiugaBachi — COMMENTED** [review]
Is this worth keeping around / modifying for logging exceptions or are `logError` splices good enough?

**Comment by jappeace** [comment]
Good work! 
Okay, so in the morrow just finish this up.
Eg make test suite pass, and make it accaptable for hacakge upload (version bounds, change log, etc), I don't think we need to do more refactoring.

File: src/Keter/App.hs [hunk 2]
```diff
@@ -133,36 +141,38 @@ data AppStartConfig = AppStartConfig
...
-                 -> AppId
+           -> (Maybe FilePath -> BundleConfig -> Maybe EpochTime -> KeterM AppStartConfig a)
+           -> KeterM AppStartConfig a
+withConfig _aid (AIData bconfig) f = f Nothing bconfig Nothing
+withConfig aid (AIBundle fp modtime) f = do
+    withRunInIO $ \rio ->
+        bracketOnError
+            (rio $ unpackBundle fp aid)
+            (\(newdir, _) -> removeDirectoryRecursive newdir)
+            $ \(newdir, bconfig) -> rio $ f (Just newdir) bconfig (Just modtime)
```
Comment (jappeace) [inline src/Keter/App.hs:157]: I'm ont sure all these changes are the result of an auto formatter, but it makes reviewiing harder if that's the case, I recomend disabling it.

**Comment by RiugaBachi** [comment]
That should do it! Test suite now passing. Let me know if anything else needs to be done.

Todo list in OP updated.

File: src/Keter/Config/V10.hs [hunk 3]
```diff
@@ -182,6 +185,7 @@ instance ParseYamlFile KeterConfig where
             <*> o .:? "missing-host-response-file"
             <*> o .:? "unknown-host-response-file"
             <*> o .:? "proxy-exception-response-file"
+            <*> o .:? "rotate-logs" .!= True
```
Comment (jappeace) [inline src/Keter/Config/V10.hs:188]: I think we should update /etc/keter-config.yaml to reflect the existence of this option. Maybe we set it to false there by default as well as emmitting to stderr is more standard unix behavior, but unset we still want to set it to true so existing users don't lose logs

File: src/Keter/Main.hs [hunk 4]
```diff
@@ -55,74 +64,93 @@ import qualified Filesystem.Path as FP (FilePath)
...
+            now <- liftIO getCurrentTime
+            -- Format: "$time|$module$:$line_num|$log_level> $msg"
+            let bs = FL.toLogStr $ encodeUtf8 $ T.pack $ concat
+                    [ take 22 $ show now
+                    , "|"
+                    , show (L.loc_module loc) <> ":" <> show (fst $ L.loc_start loc)
+                    , "|"
+                    , drop 3 $ show lvl
+                    , "> "
+                    , show msg
```
Comment (jappeace) [inline src/Keter/Main.hs:134]: calling show on msg causes it to be quoted, this is a bit reeudant:

image


keep in mind that LogStr is a monoid, 
and you can use toLogStr to convert ordinary strings to log message, no need for a Text represnetation in between.

Comment (RiugaBachi) [reply src/Keter/Main.hs:134]: I also meant to change that to a `drop 5`

Comment (jappeace) [reply src/Keter/Main.hs:134]: so to clarify, I think it'd be better to convert everything toLogStr as soon as you can and concat on that.
looks like module name is also quoted.

File: src/Keter/Config/V10.hs (hunk 3 above)
Comment (RiugaBachi) [reply src/Keter/Config/V10.hs:188]: Added

File: src/Keter/Main.hs (hunk 4 above)
Comment (RiugaBachi) [reply src/Keter/Main.hs]: Formatting should be good now.

File: src/Keter/Context.hs [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+
+import           Keter.Common
+import           Control.Monad.Trans       (lift)
+import           Control.Monad.IO.Class   (MonadIO)
+import           Control.Monad.IO.Unlift   (MonadUnliftIO)
+import           Control.Monad.Logger      (MonadLogger, MonadLoggerIO, LoggingT(..), runLoggingT)
+import           Control.Monad.Reader      (MonadReader, ReaderT, runReaderT, ask
+                                           , withReaderT)
+
+-- | The top-level keter context monad, carrying around the main logger and some locally relevant configuration structure.
```
Comment (jappeace) [inline src/Keter/Context.hs:14]: Please also refer to a relevant blogpost if you have one in mind for this design.
for example https://www.fpcomplete.com/blog/2017/06/readert-design-pattern/ 

this allows newer people to figure out what goes on with all the boilerplate :)

Comment (RiugaBachi) [reply src/Keter/Context.hs:14]: I don't think I know of one off the top of my head more specialized than that, so if you don't mind I'll steal that link lol. Added

**Review by jappeace — APPROVED** [review]
oki, I'll merge this tonight, you may fixup whatever you want meanwhile. 
other people will use this code and it's public, so, no pressure :tada:

File: src/Keter/Common.hs [hunk 6]
```diff
@@ -169,17 +68,19 @@ data KeterException = CannotParsePostgres FilePath
...
+-- logEx = do
+--     let showLoc TH.Loc { TH.loc_module = m, TH.loc_start = (l, c) } = concat
+--             [ m
+--             , ":"
+--             , show l
+--             , ":"
+--             , show c
+--             ]
+--     loc <- fmap showLoc TH.qLocation
+--     [|(. ExceptionThrown (pack $(TH.lift loc)))|]
```
[the remark this answers is not in the document]
Comment (RiugaBachi) [reply src/Keter/Common.hs]: Done

File: src/Keter/AppManager.hs [hunk 7]
```diff
@@ -211,22 +224,23 @@ performNoLock am@AppManager {..} aid action = E.mask_ $ do
                         AIData _               -> Nothing
                 tstate <- newTVar $ ASStarting Nothing tmtimestamp tmnext
                 modifyTVar apps $ Map.insert aid tstate
-                return $ launchWorker am aid tstate tmnext Nothing action
+                return $ launchWorker aid tstate tmnext Nothing action
             Terminate -> return noWorker
 
-launchWorker :: AppManager
-             -> AppId
+-- | TODO: Propagate `KeterM` up through here?
```
[the remark this answers is not in the document]
Comment (RiugaBachi) [reply src/Keter/AppManager.hs]: Done

File: src/Keter/Main.hs [hunk 8]
```diff
@@ -137,63 +163,60 @@ withManagers input mkPlugins f = withLogger input $ \kc@KeterConfig {..} log ->
...
-        AppMan.reloadAppList appMan newMap
-  where
-    incoming = getIncoming kc
+    wm <- liftIO FSN.startManager
+    withMappedConfig (const appMan) $ withRunInIO $ \rio -> do
+        _ <- FSN.watchTree wm (fromString incoming) (const True) $ \e -> do
+                e' <-
+                    case e of
+                        FSN.Removed fp _ _ -> do
+                            --log $ WatchedFile "removed" (fromFilePath fp)
```
[the remark this answers is not in the document]
Comment (RiugaBachi) [reply src/Keter/Main.hs]: Good catch! I used my fuzzy find plugin fearing this but I guess a few escaped my eyes

**Comment by jappeace** [comment]
thank you!

**Pull request merged**